### PR TITLE
Yield in try

### PR DESCRIPTION
**Pull request by yglukhov**
Includes #7704.
Adds yield in try.

### CODE REVIEW

**Comment by dom96** [comment]
I ripped out the try transformation from the ``async`` macro and ran into an internal error in the tasynctry test. Here is the branch in case you want to try, it's got the changes from this PR: https://github.com/nim-lang/Nim/tree/rip-out-try-async (you can merge it into your branch if you wish).

**Comment by dom96** [comment]
@yglukhov btw, will `yield` in `except` be supported too? :)

**Comment by yglukhov** [comment]
@dom96, of course, it is in the tests already :)

**Comment by yglukhov** [comment]
Known issues fixed. Please proceed with the review. `asyncftpclient` test failure looks unrelated.

**Comment by yglukhov** [comment]
Or it is...

**Comment by dom96** [comment]
The compiler doesn't crash anymore, but the async tests don't work with your transformation: https://github.com/nim-lang/Nim/tree/rip-out-try-async-2

**Comment by yglukhov** [comment]
@dom96, I've fixed one of the tests. The rest you should fix along with your transformation rip out.

**Comment by Araq** [comment]
So ... is this ready to be merged?

File: compiler/closureiters.nim [hunk 1]
```diff
@@ -0,0 +1,1281 @@
...
+  of nkPar, nkObjConstr, nkTupleConstr, nkBracket, nkArgList:
+    var ns = false
+    for i in 0 ..< n.len:
+      n[i] = ctx.lowerStmtListExprs(n[i], ns)
+
+    if ns:
+      needsSplit = true
+
+      result = newNodeI(nkStmtListExpr, n.info)
+      if n.typ.isNil: internalError("lowerStmtListExprs: constr typ.isNil")
```
Comment (Araq) [inline compiler/closureiters.nim]: Pretty sure ``nkArgList`` has no type.

File: compiler/closureiters.nim [hunk 2]
```diff
@@ -0,0 +1,1281 @@
...
+
+      block: # Subdivide the states
+        let oldNearestFinally = ctx.nearestFinally
+        ctx.nearestFinally = finallyIdx
+
+        let oldExcHandlingState = ctx.curExcHandlingState
+
+        ctx.curExcHandlingState = exceptIdx
+
+        discard ctx.transformReturnsInTry(tryBody)
```
Comment (Araq) [inline compiler/closureiters.nim]: These ``discards`` are concerning. The result is not to be discarded unless ``ctx.transformReturnsInTry(tryBody) == tryBody``. If so, this should be asserted.

File: compiler/closureiters.nim (hunk 1 above)
Comment (yglukhov) [reply compiler/closureiters.nim]: Ok, i just removed `nkArgList` from the case branch, since its not supposed to make it into the transformation anyway.

**Pull request merged**